base/commands/migration/migration_stages.go

```diff
@@ -0,0 +1,316 @@
+	if err != nil {
+		return nil, err
+	}
+	if it.HasNext() { // single iteration is enough that we are reading single result for a single migration
+		row, err := it.Next()
+		if err != nil {
+			return nil, err
+		}
+		r, err := row.Get(0)
+		if err != nil {
+			return nil, err
+		}
+		var status OverallMigrationStatus
+		if err = json.Unmarshal(r.(serialization.JSON), &status); err != nil {
+			return nil, err
+		}
+		for _, m := range status.Migrations {
+			dss = append(dss, DataStructureInfo{
+				Name: m.Name,
+				Type: m.Type,
+			})
+		}
+	}
+	return dss, nil
+}
+
+func saveMemberLogs(ctx context.Context, ec plug.ExecContext, ci *hazelcast.ClientInternal, migrationID string) error {
+	for _, m := range ci.OrderedMembers() {
+		l, err := ci.Client().GetList(ctx, DebugLogsListPrefix+m.UUID.String())
+		if err != nil {
+			return err
+		}
+		logs, err := l.GetAll(ctx)
+		if err != nil {
+			return err
+		}
+		for _, line := range logs {
+			ec.Logger().Info(fmt.Sprintf("[%s_%s] %s", migrationID, m.UUID.String(), line.(string)))
+		}
+	}
+	return nil
+}
+
+func saveReportToFile(ctx context.Context, ci *hazelcast.ClientInternal, migrationID, fileName string) error {
+	report, err := fetchMigrationReport(ctx, ci, migrationID)
+	if err != nil {
+		return err
+	}
+	f, err := os.Create(fmt.Sprintf(fileName))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	return os.WriteFile(fileName, []byte(report), 0600)
+}
+
+func waitForMigrationToBeCreated(ctx context.Context, ci *hazelcast.ClientInternal, migrationID string) error {
+	for {
+		status, err := fetchMigrationStatus(ctx, ci, migrationID)
+		if err != nil {
+			return err
+		}
+		if Status(status) == StatusInProgress {
+			return nil
+		}
+	}
+}
+
+type OverallMigrationStatus struct {
+	Status               Status              `json:"status"`
+	Logs                 []string            `json:"logs"`
+	Errors               []string            `json:"errors"`
+	Report               string              `json:"report"`
+	CompletionPercentage float32             `json:"completionPercentage"`
+	Migrations           []DSMigrationStatus `json:"migrations"`
+}
+
+type DataStructureInfo struct {
+	Name string
+	Type string
+}
+
+type DSMigrationStatus struct {
+	Name                 string  `json:"name"`
+	Type                 string  `json:"type"`
+	Status               Status  `json:"status"`
+	CompletionPercentage float32 `json:"completionPercentage"`
+	Error                string  `json:"error"`
+}
+
+func fetchMigrationStatus(ctx context.Context, ci *hazelcast.ClientInternal, migrationID string) (string, error) {
+	q := fmt.Sprintf(`SELECT JSON_QUERY(this, '$.status') FROM %s WHERE __key='%s'`, StatusMapName, migrationID)
+	res, err := ci.Client().SQL().Execute(ctx, q)
+	if err != nil {
+		return "", err
+	}
+	it, err := res.Iterator()
+	if err != nil {
+		return "", err
+	}
+	if it.HasNext() { // single iteration is enough that we are reading single result for a single migration
```

Comment:
https://github.com/hazelcast/hazelcast-commandline-client/pull/398/commits/ccfb42509560520de525e08f53962d059462f23b

Reply:
@kutluhanmetin Should we do the same thing for fetching migration report, dataStructuresToBeMigrated(), or fetchMigrationErrors()?

In any of these methods, if we receive some empty SqlResult, I think we can throw an error instead of silently exiting.

Reply:
Yes you are right, except for `fetchMigrationErrors`. Because if we don't have any errors, we cannot iterate. I added the same check for `dataStructuresToBeMigrated` and `fetchMigrationReport` with: https://github.com/hazelcast/hazelcast-commandline-client/pull/398/commits/2d916cf7895ecd78b179608756937c21e96f730e